### PR TITLE
Test Seeding Input Direct Seeding Defaults

### PR DESCRIPTION
**Pull request by won369369**
__Pull Request Description__

Addressing #159 

---
__Licensing Certification__

FarmData2 is a [Free Cultural Work](https://freedomdefined.org/Definition) and all accepted contributions are licensed as described in the LICENSE.md file. This requires that the contributor holds the rights to do so. By submitting this pull request __I certify that I satisfy the terms of the [Developer Certificate of Origin](https://developercertificate.org/)__ for its contents.

### CODE REVIEW

File: farmdata2/farmdata2_modules/fd2_field_kit/seedingInput/seedingInput.type.defaults.spec.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    cy.visit('/farm/fd2-field-kit/seedingInput')
+  })
+
+  afterEach(() => {
+    // Save the local storage at the end of each test so 
+    // that it can be restored at the start of the next 
+    cy.saveLocalStorage()
+  })
+
+  it('test the first visit to the page (i.e. no cached crops and areas)', () => {
```
Comment (braughtg) [inline farmdata2/farmdata2_modules/fd2_field_kit/seedingInput/seedingInput.type.defaults.spec.js:26]: This test does not appear to address anything requested in the issue and should be removed.

File: farmdata2/farmdata2_modules/fd2_field_kit/seedingInput/seedingInput.type.defaults.spec.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    expect(areas).to.equal(null)
+
+    // Set up intercepts to wait for the map to be loaded for caching. 
+    cy.intercept('GET', 'taxonomy_term?bundle=farm_crops&page=1').as('cropmap')
+    cy.intercept('GET', 'taxonomy_term.json?bundle=farm_areas').as('areamap') 
+    // Wait here for the maps to load in the page. 
+    cy.wait(['@cropmap', '@areamap']) 
+  })
+
+  //issue 159 sub-task 1
```
Comment (braughtg) [inline farmdata2/farmdata2_modules/fd2_field_kit/seedingInput/seedingInput.type.defaults.spec.js:43]: These comments are fine during development for your reference, but should be removed when complete as they will not have immediate meaning to someone reading your code in the future.

**Review by braughtg — CHANGES_REQUESTED** [review]

**Comment by braughtg** [comment]
Please review requested changes and mark the PR as ready for review again when they have been addressed.

**Comment by johnmaccormick** [comment]
You have used some advanced Cypress techniques for managing the browser's local storage, if I understand correctly. All tests are passing, so that's great. But I would like to seek the opinion of a farmdata2 maintainer about whether it is a good idea to use these advanced techniques. I'm referring to the use of

* `localStorage.removeItem()`
* `cy.restoreLocalStorage()`
* `cy.saveLocalStorage()`
* `cy.intercept()`
* `cy.wait()`

I see that Prof Braught @braughtg  already made some similar comments above.

**Comment by johnmaccormick** [comment]
I have another question about your advanced techniques. When checking the size of the list of areas, you compare it to the content of `JSON.parse(localStorage.getItem('areas'))`  -- but this raises an interesting question about testing technique. How do we know that the correct list was fetched and stored into the local storage? If we test that the fetched list has the same length as the drop-down list, have we covered the most likely type of errors intended to be covered by this test? I'm not sure of the answer, but I'm wondering if a simpler approach with a hardcoded length would be more robust. Interesting question for discussion.

**Comment by johnmaccormick** [comment]
Overall, I think you've done good work on this issue and once the file name is correct there is a good chance it can be merged. Nice job!

**Review by johnmaccormick — COMMENTED** [review]
Overall, I think you've done good work on this issue and once the file name is correct there is a good chance it can be merged. Nice job!

File: farmdata2/farmdata2_modules/fd2_field_kit/seedingInput/seedingInput.direct.defaults.spec.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    cy.get('[data-cy=num-feet-input] > [data-cy=text-input]')
+      .should('be.enabled') // check that it is enabled
+      .should('have.value', '')
+  })
+
+  //sub-task #6,7 issue 159
+  it('should display dropdown for units that is enabled', () => {
+    cy.get('[data-cy=direct-seedings]')
+      .click()
+      .then(() => {
```
Comment (braughtg) [inline farmdata2/farmdata2_modules/fd2_field_kit/seedingInput/seedingInput.direct.defaults.spec.js:55]: The assertions here should not be enclosed in a `then()`.  When inclosed in a `then()` they are executed asynchronously and this test method may exit before they execute.  This can result in any `afterEach` clause executing before the assertions.  While you do not have an `afterEach()` here, and this will likely work just fine, it is better to maintain consistency across tests by not enclosing the `should()` statements in `then()` clauses.

Note this also applies to an additional test above.

File: farmdata2/farmdata2_modules/fd2_field_kit/seedingInput/seedingInput.direct.defaults.spec.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  it('should display the Direct Seeding section when Direct is selected', () => {
+    cy.get('[data-cy=direct-seedings]').check();
+    cy.get('[data-cy=direct-area-selection]').should('be.visible');
+    cy.get('[data-cy=num-rowbed-input]').should('be.visible');
+    cy.get('[data-cy=unit-feet]').should('be.visible');
+    cy.get('[data-cy=num-feet-input]').should('be.visible');
+  });
+
+  //sub-task #2,3 issue 159
+  it('test if areas are correctly loaded to the dropdown for direct seeding', () => {
```
Comment (braughtg) [inline farmdata2/farmdata2_modules/fd2_field_kit/seedingInput/seedingInput.direct.defaults.spec.js:18]: Because the test is checking the areas, which are loaded by an API call in the page, it would be good to add a call to `cy.waitForPage()` to the start of this test.  That function call will cause the test to pause and wait for the entire page to be loaded, which will ensure that the area dropdown has been populated before the rest of the test runs.

This test should also check that the area dropdown is enabled.  See sub-task #2.

**Review by braughtg — CHANGES_REQUESTED** [review]

**Comment by braughtg** [comment]
Overall these tests are pretty good.  To be eventually be merged into the upstream there are additional issues described above that should be addressed.  Please ask here if you have questions about the requested changes.

**Comment by johnmaccormick** [comment]
Hi, I still don't see where the test checks that the area dropdown is enabled. This is subtask number 2. Am I missing something? Let me know if so, if not, please add the explicit test that checks the area dropdown is enabled. 

One more thing: As we discussed, Professor Braught asked you to remove the comments indicating the subtask numbers... But I asked you to leave them since it makes it easier for me to review. When I do merge your PR, I will manually delete these comments. To do that, I need to have write permission on your repo and also you should check the settings on your PR to make sure that maintainers can make edits. Let's see if we can get that part of the workflow operating correctly – it would be an interesting experiment. If it causes problems, I will just ask you to delete those comments yourself, but I think there is a good use case for the maintainer making some minor final tweaks before merging, as described on [StackOverflow](https://stackoverflow.com/questions/73259187/received-a-pull-request-from-someone-else-how-to-edit-it-before-merging-directl)

**Comment by won369369** [comment]
Dear Professor @johnmaccormick, 
1. We implemented the test checks that the area dropdown is enabled. 
2. What I learned from Oppia is that the blue checkbox must be visible on the PR website for maintainers to make changes to contributors' code. I don't see it from my current point of view.
image

File: farmdata2/farmdata2_modules/fd2_field_kit/seedingInput/seedingInput.direct.defaults.spec.js (hunk 4 above)
Comment (johnmaccormick) [reply farmdata2/farmdata2_modules/fd2_field_kit/seedingInput/seedingInput.direct.defaults.spec.js:25]: @won369369 @Shahir-47 @JinLeeGG  -- It looks like you still need to follow up on the suggestion to add a `cy.waitForPage()`

**Comment by johnmaccormick** [comment]
Okay, it looks like I still don't have write permission on your repo, but I'm happy to abandon that experiment. I see you have deleted the comments that help me identify the subtasks already. So, this file is nearly ready to be merged. My only remaining request is that you add a general comment at the top of the file describing in plain English what the purpose of the file is. (And also see my more detailed comment in the *Files changed* tab -- do you need to add `cy.waitForPage()`?)

**Comment by braughtg** [comment]
When the above 2 changes are addressed this will be ready for merge into the upstream.
  * Add `cy.waitForPage()`
  * Remove the sub-task comments after Prof. MacCormick confirms he no longer needs them.

**Comment by johnmaccormick** [comment]
@won369369 @JinLeeGG @Shahir-47 -- One last thing. Please add a comment at the top of the file describing what the purpose of the file is. After that, upgrade from draft to regular pull request and I think it will be ready to be merged!

**Comment by johnmaccormick** [comment]
nice job -- successfully merged
